tools/clang/test/CodeGenHLSL/batch/declarations/resources/byteaddressbuffers/store_type_shapes.hlsl

```diff
@@ -0,0 +1,46 @@
+// RUN: %dxc -E main -T vs_6_2 -HV 2018 -enable-16bit-types %s | FileCheck %s
+
+// Tests that ByteAddressBuffer.Store<T> works with all type shapes
+
+struct S
+{
+  int16_t i;
+  // 2-byte padding here, to test offsets.
+  float f;
+  struct {} _; // 0-byte field
+};
+RWByteAddressBuffer buf;
+
+void main() {
+  // CHECK: call void @dx.op.rawBufferStore.i32(i32 140, {{.*}}, i32 100, i32 undef, i32 42, i32 undef, i32 undef, i32 undef, i8 1, i32 4)
+  buf.Store(100, 42); // Test with literal int to make sure it becomes a uint, not a uint64_t
+  // CHECK: call void @dx.op.rawBufferStore.i32(i32 140, {{.*}}, i32 104, i32 undef, i32 42, i32 undef, i32 undef, i32 undef, i8 1, i32 4)
+  buf.Store(104, (int)42);
+  // CHECK: call void @dx.op.rawBufferStore.i16(i32 140, {{.*}}, i32 108, i32 undef, i16 42, i16 undef, i16 undef, i16 undef, i8 1, i32 2)
+  buf.Store(108, (int16_t)42);
+  // CHECK: call void @dx.op.rawBufferStore.i32(i32 140, {{.*}}, i32 112, i32 undef, i32 42, i32 0, i32 undef, i32 undef, i8 3, i32 8)
+  buf.Store(112, (int64_t)42);
+
+  // CHECK: call void @dx.op.rawBufferStore.i32(i32 140, {{.*}}, i32 200, i32 undef, i32 42, i32 42, i32 undef, i32 undef, i8 3, i32 4)
+  buf.Store(200, (int2)42);
+
+  // CHECK: call void @dx.op.rawBufferStore.i32(i32 140, {{.*}}, i32 300, i32 undef, i32 42, i32 42, i32 42, i32 42, i8 15, i32 4)
+  buf.Store(300, (int2x2)42);
+  // CHECK: call void @dx.op.rawBufferStore.i32(i32 140, {{.*}}, i32 316, i32 undef, i32 42, i32 42, i32 42, i32 42, i8 15, i32 4)
+  buf.Store<int2x2>(316, 42); // Also test explicit template argument and conversions
```

Comment:
could you also add test coverage for other types passed as template argument?

Reply:
If not passed explicitly, the template argument is deduced. It shouldn't depend on the type of the argument. I'll make it a separate test to make things clearer.